lib/utils/resolverHelpers.ts

```diff
@@ -0,0 +1,35 @@
+import { ClientWithEns } from "@ensdomains/ensjs/dist/types/contracts/consts";
+import { getSupportedInterfaces } from "@ensdomains/ensjs/public";
+import { Address, isAddress, PublicClient } from "viem";
+
+export const getUnsupportedResolverInterfaces = async (
+  publicClient: PublicClient & ClientWithEns,
+  resolverAddress: string
+): Promise<string[]> => {
+  if (!isAddress(resolverAddress)) {
+    return [];
+  }
+
+  const requiredInterfaces: { id: Address; name: string }[] = [
+    { id: "0x01ffc9a7", name: "ERC165" },
+    { id: "0x3b3b57de", name: "addr(bytes32)" },
```

Comment:
does this work with the `PublicResolver`, `DatabaseResolver` or the `L1Resolver`?

from my understanding, what needs to be tested is the support of the whole interface, not just a single function. first, it is only testing whether you could fetch this data, and also, the EIP-165 tests whether the contract implements the interface as a whole, which wouldn't work with the approach.

I suggest you open a Remix playground, import the ENS contracts, gather the `interfaceId` of each of their interfaces, and then check whether they are supported using this `getSupportedInterfaces`. I can help you with this.

Reply:
@pikonha It works, though some of our resolvers are not fully ENS-compliant. To fully validate the ENS interface, we would need to test each function (https://docs.ens.domains/resolvers/interfaces). This could flood our component with warnings, or we’d need to spend extra time selectively displaying only some of them.

Since a resolver that fails to implement key functions typically also lacks others, I think it’s sufficient to test only a few key ENS functions. From what I understand, the component’s goal is to warn the user if the address isn’t a valid resolver, not to verify every ENS function. What do you think?

Reply:
as far as I know, all of our resolvers are fully compliant with the ENS interfaces.

you're gonna need to add the following conditions for the offchain interfaces:

```solidity
// 0xf92709e0
interface OffchainRegister {

    /**
     * Forwards the registering of a domain to the L2 contracts
     * @param name DNS-encoded name to be registered.
     * @param owner Owner of the domain
     * @param duration duration The duration in seconds of the registration.
     * @param secret The secret to be used for the registration based on commit/reveal
     * @param resolver The address of the resolver to set for this name.
     * @param data Multicallable data bytes for setting records in the associated resolver upon reigstration.
     * @param reverseRecord Whether this name is the primary name
     * @param fuses The fuses to set for this name.
     * @param extraData any encoded additional data
     */
    function register(
        bytes calldata name,
        address owner,
        uint256 duration,
        bytes32 secret,
        address resolver,
        bytes[] calldata data,
        bool reverseRecord,
        uint16 fuses,
        bytes memory extraData
    )
        external
        payable;

}

// 0x8de0a2ed
interface OffchainRegisterParams {

    /**
     * @notice Returns the registration parameters for a given name and duration
     * @param name The DNS-encoded name to query
     * @param duration The duration in seconds for the registration
     * @return price The price of the registration in wei per second
     * @return commitTime the amount of time the commit should wait before being revealed
     * @return extraData any given structure in an ABI encoded format
     */
    function registerParams(
        bytes calldata name,
        uint256 duration
    )
        external
        view
        returns (uint256 price, uint256 commitTime, bytes memory extraData);

}

// 0xac9650d8 
interface OffchainMulticallable {

    /**
     * @notice Executes multiple calls in a single transaction.
     * @param data An array of encoded function call data.
     */
    function multicall(bytes[] calldata data)
        external
        returns (bytes[] memory);

}

```